main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -28,301 +28,346 @@ import ca.uwaterloo.flix.util.Validation._
+    case OccurrenceAst.Expression.ApplyDefTail(sym, args, tpe, loc) =>
       val as = args.map(visitExp(_, subst0))
-      Expression.ApplyDefTail(sym, as, tpe, loc)
+      LiftedAst.Expression.ApplyDefTail(sym, as, tpe, loc)
 
-    case Expression.ApplySelfTail(sym, formals, actuals, tpe, loc) =>
+    case OccurrenceAst.Expression.ApplySelfTail(sym, formals, actuals, tpe, loc) =>
       val as = actuals.map(visitExp(_, subst0))
       val fs = formals map {
-        case OccurrenceAst.FormalParam(sym, mod, tpe, loc) => FormalParam(sym, mod, tpe, loc)
+        case OccurrenceAst.FormalParam(sym, mod, tpe, loc) => LiftedAst.FormalParam(sym, mod, tpe, loc)
       }
-      Expression.ApplySelfTail(sym, fs, as, tpe, loc)
+      LiftedAst.Expression.ApplySelfTail(sym, fs, as, tpe, loc)
 
-    case Expression.Unary(sop, op, exp, tpe, loc) =>
+    case OccurrenceAst.Expression.Unary(sop, op, exp, tpe, loc) =>
       val e = visitExp(exp, subst0)
-      Expression.Unary(sop, op, e, tpe, loc)
+      LiftedAst.Expression.Unary(sop, op, e, tpe, loc)
 
-    case Expression.Binary(sop, op, exp1, exp2, tpe, loc) =>
+    case OccurrenceAst.Expression.Binary(sop, op, exp1, exp2, tpe, loc) =>
       val e1 = visitExp(exp1, subst0)
       val e2 = visitExp(exp2, subst0)
-      Expression.Binary(sop, op, e1, e2, tpe, loc)
+      LiftedAst.Expression.Binary(sop, op, e1, e2, tpe, loc)
 
-    case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+    case OccurrenceAst.Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
       val e1 = visitExp(exp1, subst0)
       val e2 = visitExp(exp2, subst0)
       val e3 = visitExp(exp3, subst0)
-      Expression.IfThenElse(e1, e2, e3, tpe, loc)
+      e1 match {
+        case LiftedAst.Expression.True(_) => e2
+        case LiftedAst.Expression.False(_) => e3
+        case _ => LiftedAst.Expression.IfThenElse(e1, e2, e3, tpe, loc)
+      }
 
-    case Expression.Branch(exp, branches, tpe, loc) =>
+    case OccurrenceAst.Expression.Branch(exp, branches, tpe, loc) =>
       val e = visitExp(exp, subst0)
       val bs = branches.map {
         case (sym, br) => sym -> visitExp(br, subst0)
       }
-      Expression.Branch(e, bs, tpe, loc)
+      LiftedAst.Expression.Branch(e, bs, tpe, loc)
 
-    case Expression.JumpTo(_, _, _) => exp0
+    case OccurrenceAst.Expression.JumpTo(sym, tpe, loc) => LiftedAst.Expression.JumpTo(sym, tpe, loc)
 
-    case Expression.Let(sym, exp1, exp2, occur, tpe, purity, loc) =>
-      /// Case 1:
-      /// If `exp1` occurs once and it is pure, then it is safe to inline without increasing neither the code size
-      /// nor the execution time.
-      val wantToPreInline = (occur, purity) match {
-        case (Occur.Once, Purity.Pure) => true
-        case _ => false
-      }
-      if (wantToPreInline) {
-        val subst1 = subst0 + (sym -> exp1)
-        visitExp(exp2, subst1)
+    case OccurrenceAst.Expression.Let(sym, exp1, exp2, occur, tpe, purity, loc) =>
```

Comment:
I think it would be nice if you had a comment that explained the cases:

///
/// Case 1: A variable is unused and its expression is pure => Remove it.
/// Case 2: ...
/// Case 2.1: ...

or whatever.

Reply:
For each case, you should answer three questions:

- Why is it safe.
- Impact on code size.
- Impact on runtime.